ribbon-eureka/src/main/java/com/netflix/loadbalancer/EurekaZoneAwareLoadBalancer.java

```diff
@@ -0,0 +1,157 @@
+ * share the same configuration at {@link com.netflix.loadbalancer.ZoneAwareLoadBalancerConfig}
+ * as they have equivalent semantic meaning.
+ *
+ * Why not re-use code? The current interface structures and implementations makes is quite difficult to
+ * decompose existing classes while maintaining good backwards compatibility.
+ *
+ * @author David Liu
+ */
+public class EurekaZoneAwareLoadBalancer extends EurekaDynamicNotificationLoadBalancer {
+
+    private ConcurrentHashMap<String, BaseLoadBalancer> balancers = new ConcurrentHashMap<String, BaseLoadBalancer>();
+
+    private static final Logger logger = LoggerFactory.getLogger(EurekaZoneAwareLoadBalancer.class);
+
+    private volatile DynamicDoubleProperty triggeringLoad;
+
+    private volatile DynamicDoubleProperty triggeringBlackoutPercentage;
+
+    void setUpServerList(List<Server> upServerList) {
+        this.upServerList = upServerList;
+    }
+
+    public EurekaZoneAwareLoadBalancer() {
+        super();
+    }
+
+    public EurekaZoneAwareLoadBalancer(IClientConfig niwsClientConfig,
+                                       Provider<EurekaClient> eurekaClientProvider,
+                                       ExecutorService serverListUpdateService) {
+        super(niwsClientConfig, eurekaClientProvider, serverListUpdateService);
+    }
+
+    @Override
+    protected void setServerListForZones(Map<String, List<Server>> zoneServersMap) {
+        super.setServerListForZones(zoneServersMap);
+        if (balancers == null) {
+            balancers = new ConcurrentHashMap<String, BaseLoadBalancer>();
+        }
+        for (Map.Entry<String, List<Server>> entry: zoneServersMap.entrySet()) {
+            String zone = entry.getKey().toLowerCase();
+            getLoadBalancer(zone).setServersList(entry.getValue());
+        }
+        // check if there is any zone that no longer has a server
+        // and set the list to empty so that the zone related metrics does not
+        // contain stale data
+        for (Map.Entry<String, BaseLoadBalancer> existingLBEntry: balancers.entrySet()) {
+            if (!zoneServersMap.keySet().contains(existingLBEntry.getKey())) {
+                existingLBEntry.getValue().setServersList(Collections.emptyList());
+            }
+        }
+    }
+
+    @Override
+    public Server chooseServer(Object key) {
+        if (!ZoneAwareLoadBalancerConfig.getEnabled() || getLoadBalancerStats().getAvailableZones().size() <= 1) {
+            logger.debug("Zone aware logic disabled or there is only one zone");
+            return super.chooseServer(key);
+        }
+        Server server = null;
+        try {
+            LoadBalancerStats lbStats = getLoadBalancerStats();
+            Map<String, ZoneSnapshot> zoneSnapshot = ZoneAvoidanceRule.createSnapshot(lbStats);
+            logger.debug("Zone snapshots: {}", zoneSnapshot);
+            if (triggeringLoad == null) {
```

Comment:
Just took this from the existing ZoneAwareLoadBalancer, not sure what you mean? It was pretty much impossible to refactor so that I can use the existing ZoneAwareLoadBalancer which is why some code is duplicated.